### PR TITLE
many fixes for ITM/SWO/SWV

### PR DESCRIPTION
**Pull request by bcantrill**
This work addresses a total of 9 issues that were preventing ITM + SWO from working on pyOCD + LPC55S69 + CMSIS-DAP.  Some of these are particular to our use case, but many are generic to pyOCD + SWO.  

1. When ITM is enabled, TPR is incorrectly modified to *disable* unprivileged writes when it intended the exact opposite, causing unprivileged stimulus writes to be silently dropped.  This work modifies TPR as intended:  to allow all unprivileged stimulus writes.  

2. The remote probe interface doesn't correctly return SWO data; the `swo_read` fails on a bad JSON encoding.  This work correctly returns a list of integers representing bytes.  

3. On CMSIS-DAP, when enabling SWO when SWO is already enabled, SWO is not first disabled, causing subsequent configuration to fail and the SWO reader to malfunction.  This work always disables SWO before configuring it.

4. On CMSIS-DAP, the size sent in a `DAP_SWO_DATA` request does not correctly adjust for the protocol overhead, potentially causing the last four bytes of the SWO packet to be left unread.  These trailing bytes will then be read on a subsequent read, causing an exception when the data payload does not constitute the expected bytes.  This work adjusts the size sent in `DAP_SWO_DATA` to reflect the protocol overhead.

5. On CMSIS-DAP, both `CommandError` and `DeviceError` are constructed without error messages, making it hard to debug failures from a stack trace.  This work modifies the construction of these errors to always specify an error message.

6. For reasons that are unclear, the SWO parser silently drops any ITM instrumentation packet for any stimulus port other than port 0.  This work modifies the SWO parser to not drop any ITM packets.  

7. The GDB server creates a separate SWO reader for each core (even though there is nothing core-specific with respect to SWO).  Because there is no locking between these threads, having multiple threads reading SWO causes more or less causes rampant corruption of the ITM stream on multicore targets.  This work creates only a single SWO reader thread, regardless of the number of cores.  

8. There is no serialization between the SWO reader thread and the GDB server.  As a result, if SWO and GDB are both simultaneously active, both the SWO reader and the GDB server are likely to experience protocol errors.  (This is likely the cause of issue #855.) This work adds a very coarse grain locking scheme between the two to assure that only one thread is talking to the underlying probe at at a time.

9. On the LPC55, resets with SWO enabled will cause ITM stimulus writes in the target to hang:  because ITM is enabled but `TRACECLKDIV` and `TRACECLKSEL` are reset, the stimulus port is not disabled but the FIFO never becomes ready.  This work explicitly sets `TRACECLKDIV` and `TRACECLKSEL` on a reset on the LPC55.

### CODE REVIEW

**Comment by flit** [comment]
Many thanks!

File: pyocd/probe/pydapaccess/cmsis_dap_core.py
Comment (flit) [inline pyocd/probe/pydapaccess/cmsis_dap_core.py:537]: The error message should be `"expected DAP_VENDOR%i" % index`.

File: pyocd/trace/swv.py
Comment (flit) [inline pyocd/trace/swv.py:52]: Fwiw, this restriction to port 0 was present because only port 0 is defined as the SWV stdio-style logging port. Instead of fixing it to port 0, we could make the port configurable.

The problem is that SWO reading and SWV processing are combined into a single class. Ultimately the intention is to migrate to a better design with a generic SWO reader thread that accepts a data flow graph for processing events, and SWV processing is done in a graph sink class.

File: pyocd/probe/tcp_probe_server.py
Comment (flit) [inline pyocd/probe/tcp_probe_server.py:415]: Should the client convert the list back into a bytearray? It's not technically necessary,  but might be nice for symmetry.

File: pyocd/gdbserver/gdbserver.py
Comment (flit) [inline pyocd/gdbserver/gdbserver.py:156]: The `DebugProbe` classes have a lock that should be used instead of creating a new one, along with `lock()` and `unlock()` methods.

**Review by flit — CHANGES_REQUESTED** [review]

**Comment by flit** [comment]
Questions/comments:

**Locking**
I'm really curious what interactions between the gdbserver and SWV thread caused issues. The existing locks seem like they should take care of it. Although, these locks were only merged in (ed54d26) on June 18, this year, and released in v0.27.0.

- All `DebugProbe` instances have a lock accessible by `lock()` and `unlock()` methods. This is used by the DP code to serialize all DP/AP register accesses.
- `MEM_AP` has its own lock for all memory transactions performed on that AP.
- `DAPAccessCMSISDAP` serializes the SWO commands with other CMSIS-DAP commands.

**Client/server**
At some point, I'd like to move to Base64 encoding for data transfers between the client and server.

**Session**
In the stuff I'm working on right now (refactoring commands and integrating with gdbserver: [refactor/commander](https://github.com/flit/pyOCD/tree/refactor/commander) in my fork) the `Session` now keeps track of all the server objects (gdbservers and probeserver). It makes sense to also track the `SWVReader` that is shared  by the gdbservers.

File: pyocd/gdbserver/gdbserver.py
Comment (flit) [inline pyocd/gdbserver/gdbserver.py:186]: There's a minor issue with limiting to core 0. The gdbserver subcommand accepts a `--core` argument so it only starts a gdbserver for the specified core(s). So if you only start up a gdbserver for core 1, SWV would not work with this check.

For now this change is fine to keep since it fixes the more common problem.

Probably the ultimate solution is to move the creation of all the console, semihosting, and SWV infrastructure to outside the `GDBServer` class.

**Comment by bcantrill** [comment]
> Questions/comments:
> 
> **Locking**
> I'm really curious what interactions between the gdbserver and SWV thread caused issues. The existing locks seem like they should take care of it. Although, these locks were only merged in ([ed54d26](https://github.com/mbedmicro/pyOCD/commit/ed54d2689410d2ba0051b56265aa41a37343d395)) on June 18, this year, and released in v0.27.0.
> 
> * All `DebugProbe` instances have a lock accessible by `lock()` and `unlock()` methods. This is used by the DP code to serialize all DP/AP register accesses.
> * `MEM_AP` has its own lock for all memory transactions performed on that AP.
> * `DAPAccessCMSISDAP` serializes the SWO commands with other CMSIS-DAP commands.

I did all of my testing on latest, for whatever it's worth. SWO commands do not, unfortunately, serialize with all other CMSIS-DAP commands:  because the locking is so fine-grained, there is nothing to prevent (say) `read_reg` from being interfered with by a `swo_read` in the (large) window between writing to and reading from the interface.

It should also be noted that this really didn't work **at all** (threads immediately die with data corruption as they start to interfere with one another) -- to the point that it's hard to believe that anyone has really tested SWO on CMSIS-DAP.  So this should be immediately reproducible for anyone connecting GDB to a target that has SWO enabled (as described by @adamgreig in #855), should you be interested in the specific failure modes.

File: pyocd/gdbserver/gdbserver.py
Comment (bcantrill) [reply pyocd/gdbserver/gdbserver.py:186]: Yes, though the right fix is to have the `SWVReader` not have any notion of core whatsoever, and to handle core-based event handling at a different layer -- but I was trying to get this working with minimal change, not rearchitect it.

File: pyocd/trace/swv.py
Comment (bcantrill) [reply pyocd/trace/swv.py:52]: The use of a particular port for a particular purpose is entirely based on convention, not definition.  Making it configurable would be fine, though I would leave that to subsequent work.

File: pyocd/probe/pydapaccess/cmsis_dap_core.py
Comment (bcantrill) [reply pyocd/probe/pydapaccess/cmsis_dap_core.py:537]: Agreed; good catch.

File: pyocd/gdbserver/gdbserver.py
Comment (bcantrill) [reply pyocd/gdbserver/gdbserver.py:156]: The `DebugProbe` class locks are too fine-grained, unfortunately.

File: pyocd/probe/tcp_probe_server.py
Comment (bcantrill) [reply pyocd/probe/tcp_probe_server.py:415]: I was trying to keep with what appeared to be the intended signature in the comments of returning a list of integers.

File: pyocd/gdbserver/gdbserver.py
Comment (flit) [reply pyocd/gdbserver/gdbserver.py:186]: Yes, exactly. And of course I don't expect you to rearchitect it! 😄

**Comment by flit** [comment]
Thanks. It did actually work at one point, but I was primarily using it without the gdbserver in custom scripts. It was partially merged upstream just so it didn't completely rot on a branch somewhere, and so hopefully someone would find it useful and help improve. 😉 Your work here is very much appreciated.

**Comment by bcantrill** [comment]
> Thanks. It did actually work at one point, but I was primarily using it without the gdbserver in custom scripts. It was partially merged upstream just so it didn't completely rot on a branch somewhere, and so hopefully someone would find it useful and help improve.  Your work here is very much appreciated.

Happy to help!  Glad that we were able to get it all working. For whatever it's worth, the multithreaded nature of pyOCD means that you definitely want to test the GDB server concurrently with the SWO reader, as that was the source of several of these issues.  But this is hard to test in any automated fashion, and definitely has a significant MCU- and probe-specific component; it's an impossibly big matrix, in many ways.

**Comment by flit** [comment]
Agreed, the complexity of testing some of the features (like RTOS support) has been an issue. For a while I've been intending to support device-specific tests, so the code under test gets verified on at least one device and the required test firmware is limited.

**Comment by flit** [comment]
Did you want to fix the "DAP_VENDOR0" log message before I merged?

**Comment by bcantrill** [comment]
> Did you want to fix the "DAP_VENDOR0" log message before I merged?

Yes, fixed that -- thanks for spotting it!

**Review by flit — APPROVED** [review]

**Pull request merged**
